lightning/src/ln/onion_message.rs

```diff
@@ -0,0 +1,96 @@
+/// Used to construct the blinded hops portion of a blinded route. These hops cannot be identified
+/// by outside observers and thus can be used to hide the identity of the recipient.
+pub struct BlindedNode {
+	/// The blinded node id of this hop in a blinded route.
+	pub blinded_node_id: PublicKey,
+	/// The encrypted payload intended for this hop in a blinded route.
+	pub encrypted_payload: Vec<u8>,
+}
+
+/// Onion messages can be sent and received to blinded routes, which serve to hide the identity of
+/// the recipient.
+pub struct BlindedRoute {
+	/// To send to a blinded route, the sender first finds a route to the unblinded
+	/// `introduction_node_id`, which can unblind its [`encrypted_payload`] to find out the onion
+	/// message's next hop and forward it along.
+	///
+	/// [`encrypted_payload`]: BlindedNode::encrypted_payload
+	pub introduction_node_id: PublicKey,
+	/// Creators of blinded routes supply the introduction node id's `blinding_point`, which the
+	/// introduction node will use in decrypting its [`encrypted_payload`] to forward the onion
+	/// message.
+	///
+	/// [`encrypted_payload`]: BlindedNode::encrypted_payload
+	pub blinding_point: PublicKey,
+	/// The blinded hops of the blinded route.
+	pub blinded_hops: Vec<BlindedNode>,
+}
+
+impl BlindedRoute {
+	/// Create a blinded route to be forwarded along `hops`.
+	pub fn new<Signer: Sign, K: Deref>(hops: Vec<PublicKey>, keys_manager: K) -> Result<Self, secp256k1::Error>
+		where K::Target: KeysInterface<Signer = Signer> {}
+}
+
+/// The final destination of an onion message.
+pub enum Destination {
+	/// We're sending this onion message to a node.
+	Node(PublicKey),
+	/// We're sending this onion message to a blinded route.
+	BlindedRoute(BlindedRoute),
+}
+
+/// A sender, receiver and forwarder of onion messages. In upcoming releases, this object will be
+/// used to retrieve invoices and fulfill invoice requests from offers.
+pub struct OnionMessager<Signer: Sign, K: Deref>
+where K::Target: KeysInterface<Signer = Signer>
+{
+	keys_manager: K,
+	// Coming soon:
+	// invoice_handler: InvoiceHandler,
+	// custom_handler: CustomHandler, // handles custom onion messages
+	pending_msg_events: Vec<MessageSendEvent>,
+}
+
+// Coming soon:
+// /// `InvoiceHandler` handles incoming invoice requests, invoices, and invoice errors.
+// pub trait InvoiceHandler {
+//   /// Handles an incoming invoice request.
+//   fn handle_invoice_request(invoice_request: InvoiceRequest) -> Result<Invoice, ()>;
+//   // Handles an incoming invoice.
+//   fn handle_invoice(invoice: Invoice) -> Result<(), ()>;
+//   // Handles an incoming invoice error.
+//   fn handle_invoice_error(invoice_error: InvoiceError);
+//   // Sends a payment.
+//   fn send_payment(..);
```

Comment:
High-level questions: What do you expect the parameters and intended use to be for this? Would `handle_invoice` call this?

Reply:
This would take `intermediate_ndoes`, a `Destination`, and an amount. It would then send an `invoice_request` onion message (OM) to the destination, who would respond with an `invoice` OM, and then pay the invoice probably using an `InvoicePayer` under the hood.